### PR TITLE
Transformer

### PR DESCRIPTION
**Pull request by szhengac**
## Description ##
This PR adds transformer model on WMT16.

### Essentials ###
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: gluonnlp/data/sampler.py [hunk 1]
```diff
@@ -106,6 +116,9 @@ class FixedBucketSampler(Sampler):
         smaller buckets.
     shuffle : bool, default False
         Whether to shuffle the batches.
+    batch_type : str, default 'sentence'
```
Comment (szha) [inline gluonnlp/data/sampler.py]: samplers don't really know what a word or sentence is. all it knows is samples. I think what's relevant here is whether to use scaleup_key or to use average bucket sample sizes. could you think of an argument name that reflects this? maybe use_bucket_avg?

File: gluonnlp/data/sampler.py [hunk 2]
```diff
@@ -194,9 +207,16 @@ def __init__(self, lengths, batch_size, num_buckets=10, bucket_keys=None,
...
-                                            * self._ratio * batch_size), batch_size)
-                                    for scale_up_key in scale_up_keys]
+        if batch_type == 'sentence':
+            self._bucket_batch_sizes = [max(int(max_scale_up_key / float(scale_up_key)
+                                                * self._ratio * batch_size), batch_size)
+                                        for scale_up_key in scale_up_keys]
+        elif batch_type == 'word':
+            bucket_average_lengths = _bucket_average_lengths(self._bucket_sample_ids, self._lengths)
+            self._bucket_batch_sizes = [max(int(batch_size / float(average_length)), 1)
+                                        for average_length in bucket_average_lengths]
```
Comment (szha) [inline gluonnlp/data/sampler.py]: the scale_up_keys are not longer used if this mode is chosen. the API doc should explain the ratio switch doesn't take effect if use_bucket_avg is true, and give warning when both are specified.

File: gluonnlp/data/translation.py [hunk 3]
```diff
@@ -194,8 +194,8 @@ def __init__(self, segment='train', src_lang='en', tgt_lang='vi',
                                         tgt_lang=tgt_lang, root=root)
 
 
-@register(segment=['train', 'newtest2012', 'newtest2013', 'newtest2014', \
-                   'newtest2015', 'newtest2016'])
+@register(segment=['train', 'newstest2012', 'newstest2013', 'newstest2014', \
```
Comment (szha) [inline gluonnlp/data/translation.py]: would you mind adding a test that catches this problem?

File: scripts/nmt/bleu.py [hunk 4]
```diff
@@ -74,6 +111,12 @@ def compute_bleu(reference_corpus_list, translation_corpus,
             'The number of translations and their references do not match'
 
     for references, translation in zip(zip(*reference_corpus_list), translation_corpus):
+        if bpe:
+            references = list(map(_bpe_to_words, references))
```
Comment (szha) [inline scripts/nmt/bleu.py]: list comprehension should be preferred in this case. e.g. `[_bpe_to_words(r) for r in references]`

File: scripts/nmt/bleu.py [hunk 5]
```diff
@@ -74,6 +111,12 @@ def compute_bleu(reference_corpus_list, translation_corpus,
...
+            translation = _bpe_to_words(translation)
+        if split_compound_word:
+            references = list(map(_split_compound_word, references))
```
Comment (szha) [inline scripts/nmt/bleu.py]: same as above

File: scripts/nmt/encoder_decoder.py [hunk 6]
```diff
@@ -584,6 +595,599 @@ def hybrid_forward(self, F, step_input, states):  #pylint: disable=arguments-dif
         return rnn_out, new_states, step_additional_outputs
 
 
+def position_encoding_init(max_length, dim):
```
Comment (szha) [inline scripts/nmt/encoder_decoder.py]: use `_` prefix for internal/hidden methods.

File: scripts/nmt/encoder_decoder.py [hunk 7]
```diff
@@ -584,6 +595,599 @@ def hybrid_forward(self, F, step_input, states):  #pylint: disable=arguments-dif
...
+        Arguments of the attention cell.
+    units : int
+    hidden_size : int
+        number of units in the hidden layer of position-wise feed-forward networks
+    num_heads : int
+        Number of heads in multi-head attention
+    scaled : bool
+        Whether to scale the softmax input by the sqrt of the input dimension in multi-head attention
+    dropout : float
+    use_residual : bool
```
Comment (szha) [inline scripts/nmt/encoder_decoder.py]: nit: missing docs. maybe explain where the dropout and residual connection is used?

File: scripts/nmt/encoder_decoder.py [hunk 8]
```diff
@@ -584,6 +595,599 @@ def hybrid_forward(self, F, step_input, states):  #pylint: disable=arguments-dif
...
+        outputs = self.layer_norm(outputs)
+        return outputs
+
+
+class TransformerEncoderCell(HybridBlock):
+    """Structure of the Transformer Encoder Cell.
+
+    Parameters
+    ----------
+    attention_cell : AttentionCell or str, default 'multi_head'
```
Comment (szha) [inline scripts/nmt/encoder_decoder.py]: explain what options are there by enumerating them. e.g. `{'multi_head', 'scaled_luong' ...}`

File: scripts/nmt/encoder_decoder.py [hunk 9]
```diff
@@ -629,3 +1233,58 @@ def get_gnmt_encoder_decoder(cell_type='lstm', attention_cell='scaled_luong', nu
...
+    units : int
+    hidden_size : int
+    dropout : float
+    use_residual : bool
+    max_src_length : int
+    max_tgt_length : int
+    weight_initializer : mx.init.Initializer or None
+    bias_initializer : mx.init.Initializer or None
+    prefix :
+    params :
```
Comment (szha) [inline scripts/nmt/encoder_decoder.py]: Need to explain the parameters.

File: scripts/nmt/encoder_decoder.py [hunk 10]
```diff
@@ -584,6 +595,599 @@ def hybrid_forward(self, F, step_input, states):  #pylint: disable=arguments-dif
...
+        self._units = units
+        self._use_residual = use_residual
+        self.ffn_1 = nn.Dense(units=hidden_size, flatten=False, activation=activation,
+                              weight_initializer=weight_initializer, bias_initializer=bias_initializer,
+                              prefix='ffn_1_')
+        self.ffn_2 = nn.Dense(units=units, flatten=False,
+                              weight_initializer=weight_initializer, bias_initializer=bias_initializer,
+                              prefix='ffn_2_')
+        self.dropout_layer = nn.Dropout(dropout)
+        self.layer_norm = nn.LayerNorm()
```
Comment (sxjscience) [inline scripts/nmt/encoder_decoder.py]: We'd better add a prefix for layernorm. Also, do we need to use `with self.name_scope():`?

File: scripts/nmt/loss.py [hunk 11]
```diff
@@ -43,9 +45,99 @@ def hybrid_forward(self, F, pred, label, valid_length):
...
+        if self._sparse_label:
+            assert units is not None or self._units is not None, \
+                'units needs to be given in function call or ' \
+                'instance initialization when sparse_label is False'
+            if units is None:
+                units = self._units
+            inputs = F.one_hot(inputs, depth=units)
+        if units is None and self._units is None:
+            return F.Custom(inputs, epsilon=self._epsilon, axis=self._axis,
+                            op_type='_smoothing_with_dim')
```
Comment (sxjscience) [inline scripts/nmt/loss.py]: We need to think how to avoid the CustomOp.

File: gluonnlp/data/translation.py [hunk 12]
```diff
@@ -253,3 +253,67 @@ def __init__(self, segment='train', src_lang='en', tgt_lang='de',
...
                                          tgt_lang=tgt_lang,
                                          root=os.path.join(root, _get_pair_key(src_lang, tgt_lang)))
+
+@register(segment=['train', 'newstest2009', 'newstest2010', 'newstest2011', \
+                   'newstest2012', 'newstest2013', 'newstest2014'])
+class WMT2014BPE(_TranslationDataset):
+    """Preprocessed Translation Corpus of the WMT2014 Evaluation Campaign.
+
+    We use the preprocessing script in
+    /scripts/nmt/wmt14_en_de.sh
```
Comment (sxjscience) [inline gluonnlp/data/translation.py]: Have we uploaded this file?

Comment (szha) [reply gluonnlp/data/translation.py]: yup

File: scripts/nmt/bleu.py (hunk 4 above)
Comment (szhengac) [reply scripts/nmt/bleu.py]: I think list(map()) is faster?

File: scripts/tests/test_encoder_decoder.py [hunk 13]
```diff
@@ -1,7 +1,8 @@
 import numpy as np
 import mxnet as mx
 from mxnet.test_utils import assert_almost_equal
-from ..nmt.encoder_decoder import *
+from ..nmt.gnmt_encoder_decoder import *
+from ..nmt.transformer_encoder_decoder import *
```
Comment (szha) [inline scripts/tests/test_encoder_decoder.py]: let's have the following module names:
nmt.encoder_decoder: defines the encoder decoder and shared functionality
nmt.gnmt: everything in gnmt including the encoder and decoder
nmt.transformer: everything in transformer.

Comment (szhengac) [reply scripts/tests/test_encoder_decoder.py]: How about the training scripts? we have gnmt.py and transformer.py

Comment (szhengac) [reply scripts/tests/test_encoder_decoder.py]: I was considering using train_gnmt.py and train_transformer.py. However, our other scripts do not use 'train' in the script name. So I keep using these to ensure consistency.

Comment (szha) [reply scripts/tests/test_encoder_decoder.py]: train_ prefix sounds good

File: scripts/nmt/loss.py (hunk 11 above)
Comment (szhengac) [reply scripts/nmt/loss.py]: Currently, it will not go to the CustomOp if the units is provided.

**Comment by mli** [comment]
Job PR-133/25 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-133/25/index.html

**Comment by mli** [comment]
Job PR-133/26 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-133/26/index.html

**Comment by mli** [comment]
Job PR-133/27 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-133/27/index.html
